test/e2e/annotations/canary.go

```diff
@@ -192,4 +192,47 @@ var _ = framework.IngressNginxDescribe("Annotations - canary", func() {
 			Expect(body).ShouldNot(ContainSubstring("http-svc-canary"))
 		})
 	})
+
+	Context("Single canary Ingress", func() {
+		It("should not use canary as a catch-all server", func() {
+			host := "foo"
+			canaryIngName := fmt.Sprintf("%v-canary", host)
+			annotations := map[string]string{
+				"nginx.ingress.kubernetes.io/canary":           "true",
+				"nginx.ingress.kubernetes.io/canary-by-header": "CanaryByHeader",
+			}
+
+			ing := framework.NewSingleIngress(canaryIngName, "/", "", f.IngressController.Namespace, "http-svc-canary", 80, &annotations)
+			f.EnsureIngress(ing)
+
+			f.WaitForNginxServer("_",
+				func(server string) bool {
```

Comment:
you can do

```
upstreamName := fmt.Sprintf(`set $proxy_upstream_name "%s-%s-%s";`, f.IngressController.Namespace, "http-svc-canary", "80")
```
...
```
return Expect(server).Should(ContainSubstring(`set $proxy_upstream_name "upstream-default-backend";`)) &&
  Expect(server).ShouldNot(ContainSubstring(upstreamName))
```

Asserting on ingress name is not necessary. I also asserting on `_` is not necessary because `WaitForNginxServer("_",` takes care of it.

Reply:
why would you check that upstream name is `upstream-default-backend` and not something else at the same time?

Isn't `Expect(server).Should(ContainSubstring(`set $proxy_upstream_name "upstream-default-backend";`))` enought?

Reply:
@wayt imagine the controller creates following lines in the Nginx configuration by mistake:

```
set $proxy_upstream_name "upstream-default-backend";
...
set $proxy_upstream_name "<upstream name for that canary app>";
```

then your test would still pass, but Nginx would be proxying to the canary app.

Reply:
Alright 👍